### PR TITLE
perf: remap earcut indices and unsubdivided fill outlines through a per-vertex lookup

### PR DESCRIPTION
**Pull request by cherenkov**
## Launch Checklist

`Subdivider` deduplicates input vertices through a `Map`. Two places then looked those same vertices up in that `Map` all over again:

- `_convertIndices` re-hashed every corner of every earcut triangle.
- `_generateOutline` re-hashed every outline point through `_pointArrayToIndices`. Below granularity 2 — every mercator tile — `subdivideVertexLine` only copies the ring and closes it, so those points are exactly the input vertices.

This records each input vertex's deduplicated index once, where the vertices are added, and reads both sites from that array. Earcut's indices are remapped in place, so `_convertIndices` is gone; an unsubdivided ring reads its own slice. `_vertexToIndex` also does one `Map` lookup instead of `has` + `get`, and the ring-closing rule is shared with `subdivideVertexLine` rather than written twice.

At granularity 2 and above the ring really is subdivided, and that path is unchanged.

Output is identical: vertex, triangle and outline counts match on the bench fixture tiles, and the fill render tests pass.

| Measurement | Before | After |
|---|---|---|
| `subdivision.bench.ts`, without subdivision | 4.25 ms | 3.69 ms |
| `subdivision.bench.ts`, granularity 64 | 82.2 ms | 78.5 ms (within noise, as expected - this path is untouched) |
| `worker_tile.bench.ts` | 19.0 ms | 18.4 ms |
| Worker CPU profile, 3 dense city loads | `_convertIndices` 46 ms + `_generateOutline` 21 ms | both gone |

For scale, on the trace in #8426 this is worth roughly 2% of wall clock, not 10-12% - the workers are not where the time goes.

 - [x] Confirm **your changes do not include backports from Mapbox projects**
 - [x] Briefly describe the changes in this PR.
 - [x] Write tests for all new functionality (covered by the existing `subdivision.test.ts`, which asserts outline and mesh integrity at granularity 0).
 - [x] Post before/after bench results.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.
 - [x] Confirm you have read our AI policy.

### CODE REVIEW

**Comment by HarelM** [comment]
It's not clear what this improves, can you clarify please? 
It's hard to understand the change, is this class fully covered in terms of coverage? Since this looks like a refactoring of a single class it might good to make sure the input and output are covered before changing it. 
Also it might be useful to make sure we improve performance where it has significant value by making sure we have a good idea where these places are, I don't feel these PRs are "planned" that way right now. 
I truly value the work on this nevertheless.

**Comment by cherenkov** [comment]
Thanks for the review - the description was too thin, sorry.

**What it improves.** Each input vertex already gets a deduplicated index, but `_convertIndices` re-looked-up every earcut triangle corner in the `Map`, and at `granularity < 2` (every mercator tile) so did `_generateOutline`. Recording that index once removes 46 ms + 21 ms in a worker profile over three dense city loads (~10-12% of worker busy time), with identical output.

**Coverage.** `subdivision.test.ts` alone covers this file at 93.8% statements / 100% functions, and its duplicate-vertex and duplicate-edge cases hit exactly the path I changed.

**On planning.** Agreed. Profiling liberty with city jumps at z13.5, the main thread is the real bottleneck (2089 ms busy vs 609 ms worker, CJK glyph SDF alone 27.6%); I took this one first only because it's small and output-identical. I'll open an issue with the profile and a ranked list.

**Comment by cherenkov** [comment]
Opened #8426 with the profile and a ranked list, as promised.

One correction to my own numbers above, now that the whole profile is written up. The 46 ms + 21 ms is 10-12% of *worker* busy time, but the workers are not where the time goes: 609 ms worker against 2089 ms on the main thread. So on that trace this PR is worth roughly 2% of the wall clock, not 10-12%. I would rather say that plainly than let it look bigger than it is.

What I think still makes it worth taking: the output is identical (vertex, triangle and outline counts match on the bench fixtures, fill render tests pass), it removes work rather than trading one cost for another, and `_convertIndices` and `_generateOutline` end up reading as what they actually do.

That said, you were right that these were not planned around where the time is. If you would rather spend the review budget on the main-thread items in #8426, I am happy to close this one.

File: src/render/subdivision.ts [hunk 1]
```diff
@@ -765,9 +782,7 @@ export function subdivideVertexLine(linePoints: Point[], granularity: number, is
 
     // Generate an extra line segment between the input array's first and last points,
     // but only if isRing=true AND the first and last points actually differ.
-    const first = linePoints[0];
-    const last = linePoints[linePoints.length - 1];
-    const addLastToFirstSegment = isRing && (first.x !== last.x || first.y !== last.y);
+    const addLastToFirstSegment = isRing && ringIsOpen(linePoints);
```
Comment (HarelM) [inline src/render/subdivision.ts:767]: I think that's not exactly as before, right? This now uses a reference and not an actual check of numbers...?

File: src/render/subdivision.ts [hunk 2]
```diff
@@ -34,6 +34,12 @@ class Subdivider {
      * Map of "vertex x and y coordinate" to "index of such vertex".
      */
     private _vertexDictionary: Map<number, number> = new Map<number, number>();
+
+    /**
+     * Index into the vertex buffer for each input vertex, in the order `flatten` produced them.
+     */
+    private _inputRemap: number[] = [];
```
Comment (HarelM) [inline src/render/subdivision.ts]: Is it possible to keep this "in memory" instead of as part of the class' state?

File: src/render/subdivision.ts [hunk 3]
```diff
@@ -418,11 +425,13 @@ class Subdivider {
...
      */
-    private _generateOutline(polygon: Point[][]): number[][] {
+    private _generateOutline(polygon: Point[][], ringStarts: number[]): number[][] {
         const subdividedLines: number[][] = [];
-        for (const ring of polygon) {
-            const line = subdivideVertexLine(ring, this._granularity, true);
-            const pathIndices = this._pointArrayToIndices(line);
+        for (let r = 0; r < polygon.length; r++) {
+            const ring = polygon[r];
+            const pathIndices = this._granularity < 2 ?
```
Comment (HarelM) [inline src/render/subdivision.ts:427]: Is this a different optimization for _granularity < 2 ? I didn't see it in any of the comments. Did I miss it?
Not that I mind this optimization, but the following tsdoc comment is not descriptive:
```
Path indices of an unsubdivided ring, read from `_inputRemap`; empty and closed as `subdivideVertexLine` would make it.
```

File: src/render/subdivision.ts (hunk 1 above)
Comment (cherenkov) [reply src/render/subdivision.ts:767]: `Point.equals` compares coordinates, not identity — in `@mapbox/point-geometry` it is:

```js
equals(other) {
    return this.x === other.x &&
           this.y === other.y;
}
```

So `!ring[0].equals(ring[ring.length - 1])` expands to exactly the old `first.x !== last.x || first.y !== last.y`. No behaviour change.

It moved into a helper because `_ringPathIndices` needs the same "does this ring need a closing segment?" rule, and I did not want the two copies to drift.

Comment (HarelM) [reply src/render/subdivision.ts:767]: Oh, that's fine then.

File: src/render/subdivision.ts (hunk 2 above)
Comment (cherenkov) [reply src/render/subdivision.ts]: Done — the field is gone; `_initializeVertices` returns the remap and it lives as a local.

That also let `flatten` go back untouched (ring starts are just `ring.length` accumulated) and `_convertIndices` disappear — earcut's array is used nowhere else, so the remap happens in place.

File: src/render/subdivision.ts (hunk 3 above)
Comment (cherenkov) [reply src/render/subdivision.ts:427]: You did not miss it, it was buried in a parenthesis in the description. Rewritten.

Same optimization, second site: `_convertIndices` re-looked-up every triangle corner, `_generateOutline` every outline point. At `granularity < 2` — every mercator tile — `subdivideVertexLine` only copies and closes the ring, so those points are the input vertices and the remap already has their indices.

Agreed on the tsdoc:

```
The outline path `subdivideVertexLine(ring, granularity, true)` returns when the granularity is too low to
subdivide, as vertex buffer indices rather than points, read from `inputRemap` instead of looked up per point.
```

File: src/render/subdivision.ts [hunk 4]
```diff
@@ -587,23 +606,24 @@ class Subdivider {
...
 
         // Subdivide triangles
         let subdividedTriangles: number[];
         try {
-            // At this point this._finalVertices is just flattened polygon points
             const earcutResult = earcut(flattened, holeIndices);
-            const cut = this._convertIndices(flattened, earcutResult);
-            subdividedTriangles = this._subdivideTrianglesScanline(cut);
+            for (let i = 0; i < earcutResult.length; i++) {
+                earcutResult[i] = inputRemap[earcutResult[i]];
```
Comment (HarelM) [inline src/render/subdivision.ts]: I'm not too excited about overriding the results from earcut. Can the `inputRemap` be passed to the method so that the earcut results are more "immutable"?

Comment (cherenkov) [reply src/render/subdivision.ts]: Fixed in 718200a2689c6cb387a5cea7fe478b67bf67afb4: earcut's output is no longer overwritten. The remapped indices go into a new array at the call site, which is the same single allocation `_convertIndices` used to make, and it benches the same as passing `inputRemap` into the method (3.61 vs 3.63 ms, rme < 1%).

I kept it out of `_subdivideTrianglesScanline` because passing it in changes what `inputIndices` means there, and the `granularity < 2` branch would then have to forward the remap into the exported `fixWindingOrder` as well. Let me know if you'd rather have it passed in.

File: src/render/subdivision.ts [hunk 5]
```diff
@@ -587,23 +606,21 @@ class Subdivider {
...
+        const inputRemap = this._initializeVertices(flattened);
 
         // Subdivide triangles
         let subdividedTriangles: number[];
         try {
-            // At this point this._finalVertices is just flattened polygon points
             const earcutResult = earcut(flattened, holeIndices);
-            const cut = this._convertIndices(flattened, earcutResult);
-            subdividedTriangles = this._subdivideTrianglesScanline(cut);
+            subdividedTriangles = this._subdivideTrianglesScanline(remapIndices(earcutResult, inputRemap));
```
Comment (HarelM) [inline src/render/subdivision.ts]: Does it make sense to move the remap into the subdivede method?

**Comment by HarelM** [comment]
Mostly looks good now, added a last nit-picking comment.

File: src/render/subdivision.ts (hunk 5 above)
Comment (cherenkov) [reply src/render/subdivision.ts]: Moved in aa0d46fea. `_subdivideTrianglesScanline` now takes `inputRemap`, so earcut's output goes in untouched. Only the `granularity < 2` branch materialises a remapped array; the subdividing path applies the remap while building `triangleIndices`, so that allocation is gone.

Slightly faster as a result: granularity 64 averaged 69.8 ms before and 67.4 ms after over five interleaved runs, though per-run rme is ±1-2%, so treat it as a couple of percent. `without subdivision` is unchanged.

**Review by HarelM — APPROVED** [review]
